sig-multicluster/namespace-sameness-position-statement.md

```diff
@@ -0,0 +1,35 @@
+# Namespace Sameness - SIG Multicluster Position Statement
+
+Author: Jeremy Olmsted-Thompson (**[@jeremyot](https://github.com/jeremyot)**), Google  
+Last Edit: 2020/03/24  
+Status: IN REVIEW  
+
+## Goal
+To establish a normative statement for multi-cluster namespace semantics and
+governance as a building block for further development which will require
+specifying behaviors across clusters.
+
+## Context
+Users are reaching for multi-cluster deployments for a
+[variety of reasons](http://bit.ly/k8s-multicluster-conversation-starter-doc).
+However, Kubernetes treats the cluster boundary as the edge of the universe.
+There are currently no standard practices for how to extend the Kubernetes
+resource model across multiple clusters. Without common patterns we can’t build
+portable tooling to facilitate multi-cluster capabilities and know that behavior
+will be consistent for each user.
+
+## Scope
+A single organization may need multiple, disjoint sets of clusters. They may,
+for example, represent different phases of the development lifecycle (dev,
+staging, prod) or support unrelated projects. Each organization governs their
+own clusters in isolation, so the scope of a namespace can only reasonably be
+declared within the organizational boundary. The scope of namespace identity is
+defined as the union of clusters, governed by a single authority, that are
+expected to work together.  An authority is a company, organization, team,
+individual, or other entity which is entrusted to manage these clusters and, in
+particular, to create namespaces in them.
+
+## Position
+**For a set of related clusters governed by a single authority, all namespaces of
```

Comment:
Can a single cluster be part of multiple authorities for disjoint subsets of namespaces.  For instance, say you have clusters 1-10.  A single ops team is responsible for all 10 clusters and wants to provide some common policies for them and a few namespaces to enforce them.  That single ops team then assigns cluster 1-5 to new-stuff app team and assigns clusteres 6-10 to legacy-stuff app team.

In this scenario, you can see three levels of authority.
 1. the ops team, which owns some namespaces in all 10 clusters
 2. the new-stuff-app-team which owns some namespaces in clusters 1-5
 3. the legacy-stuff-app-team which owns some namespaces in clusters 6-10

You can imagine this as a case where clusters themselves are managed by some external entity (cloud provider?), a corporate level policy enforcement, and then the app teams.  This position doesn't make it clear whether such a model is being explicitly excluded.

Reply:
The way I think about this would be a bit different.

Your ops team owns all 10 clusters, up to and including creation of Namespaces in them.  Through usual mechanisms you would grant the teams access to their respective namespaces in whatever clusters they are meant to have access to.

It's a little vague without an actual group implementation, but maybe your grouping allows labels on clusters (seems reasonable) and you have a centralized RBAC syncer that reads a config file or git repo or LDAP  or something.  It syncs RBAC rules to allow "new-stuff-app-team" into clusters labelled "primary=new-stuff" and syncs rules to allow "legacy-stuff-app-team" into clusters labelle "primary=legacy-stuff".  As long as those sets of namespaces are disjoint, there should be zero problem.

Neither team can willy-nilly make new namespaces in their clusters, because that could violate sameness.  But maybe you allow them to make namespaces that are prefixed by the value of their "primary", so `kubectl create ns foo` fails but `kubectl create ns new-stuff-foo` is allowed.  Obviously you need some admission that enforces this... which I see as an opportunity to start building cool new policy enforcement primitives :)

Reply:
We know of systems in the wild that make use of namespaces in local clusters to manage those clusters.  In fact, in many ways the idea of operators encourages a pattern of local control where a local namespace can be used to force local control on a cluster.  Even to the point of local agents being present on these clusters.

In the example I provided you can see parallels to the gke-connect-agent or the IBM klusterlet.  Both of which end up in namespaces on the spoke clusters.  I think it's hard to reconcile an approach that is good for google, good for IBM, (probably good for vmware), but not good for standard cluster-admins.

Reply:
> Neither team can willy-nilly make new namespaces in their clusters, because that could violate sameness. But maybe you allow them to make namespaces that are prefixed by the value of their "primary", so `kubectl create ns foo` fails but `kubectl create ns new-stuff-foo` is allowed. Obviously you need some admission that enforces this... which I see as an opportunity to start building cool new policy enforcement primitives :)

I think this may be agreeing that the use-case is valid as long as the namespaces that each group's namespaces are disjoint.  An invariant that is hard to enforce, but as a position makes sense to me.

If so, can we clarify that it is acceptable for a spoke cluster to be associated with multiple "hubs" or "superclusters" (I prefer hub), where each hub attempts to manage a disjoint set of namespaces (the details of how are left as an exercise to the reader).

I think this subtle change in position has impacts in downstream decisions about unioning services as a for instance.

Reply:
It is common for clusters to have some segmentation of namespaces where each segment may be managed by different authorities oriented around particular use cases.  It is also common to establish a means to exclude particular namespaces from management by a particular authority (either by label, name, or prefix) for that use case.  For example, a particular agent may own forwarding of metrics/logs to some centralized location off cluster, and it may exclude namespace content by label, name, prefix, etc.

I think the key distinction for me is clarifying that an authority may govern sets of clusters, but subsets of namespaces, and in that subset, its reasonable to assert that a given name is considered to be the equivalent namespace by that authority.  It is perfectly ok to have local namespaces (that may have a common name), but are excluded by that governing authority for legitimate reasons.

Reply:
In my mental model of how this fits together, here's how I can see handling the above situations:

The authority that governs a set of clusters also governs all namespaces within that set, but may delegate ownership of subsets of namespaces. The authority could assign governance of all namespaces prefixed with e.g. `new-stuff-` to the new stuff app team who can create, subdivide, or restrict namespaces in their prefixed subset as they see fit. As Tim said, it's fine if `new-stuff-` namespaces are only allowed in clusters 1-5. The meaning and ownership of the namespace is still the same across all 10 clusters even though 6-10 don't contribute.

Per @deads2k's example, the new stuff app team could manage their namespaces across clusters as they see fit, associated with their own supercluster/hub, so long as all cross-cluster operations are confined to their `new-stuff-` namespaces.

These situations can all work with the same-namespace/single-owner model so long as namespace subsets are always consistent across all clusters in the set. 

Alternatively as a bit of a special case, the authority could decide that a subset of namespaces (say prefixed with `local-`, or e.g. `default`) are purely cluster-local and have no relation with same-named namespaces in other clusters. This would be a break from the sameness principle and mean that any future development building on the assumption of cross-cluster sameness would not apply to those namespaces. To me, that seems completely reasonable as long as the cluster-authority consciously makes that decision.

Reply:
I started a short answer and it turned into a book.  Sorry.  If you want to talk on zoom or something, let me know.  I hope this makes it clearer, but I fear the nuance is hard to express in text.

@deads2k

> We know of systems in the wild that make use of namespaces in local clusters
> to manage those clusters. In fact, in many ways the idea of operators
> encourages a pattern of local control where a local namespace can be used to
> force local control on a cluster. Even to the point of local agents being
> present on these clusters.

I hope I didn't imply that this is not feasible.  Only that namepsace names
matter, and should be considered when taking such actions.

> In the example I provided you can see parallels to the gke-connect-agent or
> the IBM klusterlet. Both of which end up in namespaces on the spoke clusters.

Sure.  And if you run your agent in the same-named namespace in each cluster,
they are considered "the same".  That seems like what you would want, to me.
Again, considering an RBAC syncer, you would want the overall infra-ops group
to have the same access to that agent, regardless of which cluster it is in.

As an aside - operators are funny things.  Containers were, in part, created so
that developers did not have to bother their sysadmins to get dependencies
installed.  It became self-service.  If I need mysql version X, I could have
it, but I HAD TO OPERATE IT.  Now operators take away that freedom, and
developers have to bother cluster-admins to get their dependencies installed.
Until we address that, we should not think of CRDs and admission controllers
and operators as developer-owned things.  They are infrastructure.

> I think it's hard to reconcile an approach that is good for google, good for
> IBM, (probably good for vmware), but not good for standard cluster-admins.

I don't think this is "not good", it's just asking multi-cluster environments to
put some thought into the relationships between clusters.  If you don't want
any relationship, then you can consider every cluster it's own "group".  But as
features like multi-cluster services and multi-cluster deployments take root,
those relationships become meaningful.  I hope.

> If so, can we clarify that it is acceptable for a spoke cluster to be
> associated with multiple "hubs" or "superclusters" (I prefer hub), where each
> hub attempts to manage a disjoint set of namespaces (the details of how are
> left as an exercise to the reader).

We avoided the word "hub" because we have a product called "hub" and didn't
want people conflating the two concepts or thinking that we were trying to
force-couple them.  Quite the opposite, we want to make sure Kubernetes DOESN'T
couple, by defining the assumptions clearly. :)

As for multiple super-clusters, I worry about the ability to actually use that.
having multiple non-coordinated owners of a cluster seems like it will run into
trouble with the number of cluster-scoped things we have.  We already talked
about how prefixes could be used to uniquify namespace names, but that's maybe
the easy one.  Concretely, things like CRDs are a great example, but also DNS.
If we implement multi-cluster services, there's almost certainly a DNS tie-in.
That brings a lot of complexity to both DNS serving and kubelet to change the
DNS config based on which cluster and namespace (and thereby super-cluster) a
pod is in,

> I think this subtle change in position has impacts in downstream decisions
> about unioning services as a for instance.

I agree!  This position statement is pretty carefully scoped to NOT define how
a Services implementation *must* work.  As written, I think it leaves room for
a multi-super-cluster implementation under the same "authority", but with
different namespace names.  I'll respond to Derek and show an example of what I
mean, and you tell me if it gets close to what you're thinking.

@derekwaynecarr

> I think the key distinction for me is clarifying that an authority may govern
> sets of clusters, but subsets of namespaces, and in that subset, its
> reasonable to assert that a given name is considered to be the equivalent
> namespace by that authority. It is perfectly ok to have local namespaces
> (that may have a common name), but are excluded by that governing authority
> for legitimate reasons.

I think there are 2 things in this statement.

First, "an authority may govern sets of clusters, but subsets of namespaces".
This is missing the point, so maybe we need more words to clarify.  The
authority owns the cluster and weilds the sole power to create namespaces,
CRDs, etc. and to delegate said power to other actors.

Second, "It is perfectly ok to have local namespaces (that may have a common
name), but are excluded by that governing authority for legitimate reasons".
The authority is not excluded, but might not be taking any particular action.
Merely being the same does not necessarily activate anything, but should you
choose to activate a feature (e.g. multi-cluster service) the name-sameness
would matter.

Let's get hypothetical.

I have two clusters, "A" and "B".  I have a cluster-ops team which is responsible for
keeping them alive.  I also have app-teams, "foo" and "bar" which use those
clusters.

What I think makes sense:

Cluster-ops owns the creation of namespaces in both clusters.  They might
delegate (through a self-service portal or tooling or admission control) the
ability to create namespaces to foo-team and bar-team, but any namespace name
that is allocated in "reserved" in all clusters.

If foo-team requests a namespace called "database" in cluster A, no other team
may request a namespace called "database" in cluster B.  That name is taken.

If foo-team wants a mysql operator in cluster A, and bar-team wants a mysql
operator in cluster B, cluster-ops team should consider if those are "the same"
(e.g. the same versions, and the set of people administer it in both clusters)
or not.  Absent any other multi-cluster capabilities, it doesn't actually
matter much.  But if they have, for example, an RBAC sync-from-LDAP process, it
would probably assume that same namespace => same RBAC.  This is assuming that
LDAP->RBAC was on-by-default, which may or may not be true.

If cluster-ops has a metrics service they run in each cluster, they should
consider if those are "the same" or not (for the same reasons as above).  If
they have a multi-cluster services feature, it would probably assume that same
namespace => same service.  This is assuming that service-merging was
on-by-default, which may or may not be true (it's NOT true in the current
proposal - you must opt-in to merging, but it assumes namespace sameness).

If you really wanted to have multiple super-clusters, you could configure that.
Your LDAP->RBAC syncer and services merger and so forth would all need to be
aware of the fact that there are N configurations, but namespace sameness still
applies.

What I think DOES NOT make sense:

Clusters A and B both have a "database" namespace, but in cluster A that is
part of super-cluster "foo" and in cluster B that is part of super-cluster
"bar".

I think that makes things very confusing, and makes mistakes much more likely.
It would require more explicit configuration (as opposed to the convention of
namespace sameness).  It would be vastly more obvious if cluster A had
namespace "foo-database" and cluster B had namespace "bar-database" (my kingdom
for one extra nesting level!).

Lastly, nothing precludes the multi-cluster feature implementations for getting
crazy.  If your LDAP->RBAC syncer wants to have specializations for particular
clusters, that's great.  Maybe all clusters in EU have more restrictive RBAC or
something.  But the PURPOSE of the namespace can be assumed to be the same
across clusters, by virtue of namespace sameness.